# Sorting Data with BigQuery Public Dataset - CDC Births Data Summary

In this notebook, we’ll sort data by using SQL queries with ORDER BY and WHERE clauses. This will enable us to organize and use data more efficiently as a data analyst.

## Dataset
We’ll query the <b>CDC Births Data Summary public dataset</b>. \
We will write queries to obtain some information about which counties in the United States have the most and least births in the years 2016-2018. 

## Load the dataset

In [1]:
from google.cloud import bigquery

print(bigquery.__version__)

3.40.1


In [2]:
client = bigquery.Client()

print(client.project)

myproject001-504709


In [3]:
# This code only tests if the BigQuery connection is on.
query = """
SELECT
  name,
  SUM(number) AS total_births
FROM `bigquery-public-data.usa_names.usa_1910_current`
GROUP BY name
ORDER BY total_births DESC
LIMIT 10
"""

df = client.query(query).to_dataframe()

df.head()

,name,total_births
0,James,5054074
1,John,4910976
2,Robert,4763102
3,Michael,4396482
4,William,3939912


In [4]:
# Load the data
query = """
SELECT 
  * 
FROM 
  `bigquery-public-data.sdoh_cdc_wonder_natality.county_natality` 
LIMIT 
  1000
"""

data = client.query(query).to_dataframe()
data.head()

,Year,County_of_Residence,County_of_Residence_FIPS,Births,Ave_Age_of_Mother,Ave_OE_Gestational_Age_Wks,Ave_LMP_Gestational_Age_Wks,Ave_Birth_Weight_gms,Ave_Pre_pregnancy_BMI,Ave_Number_of_Prenatal_Wks
0,2018-01-01,"Calhoun County, AL",01015,1265,26.67,38.11,38.25,3168.14,28.85,10.45
1,2018-01-01,"Tulsa County, OK",40143,8933,28.13,38.18,38.25,3243.39,27.58,11.04
2,2018-01-01,"Carroll County, GA",13045,1540,27.20,38.50,38.50,3253.35,27.98,10.75
3,2018-01-01,"Saginaw County, MI",26145,2182,27.41,38.75,38.50,3226.90,28.85,11.58
4,2018-01-01,"Hillsborough County, FL",12057,17126,28.97,38.41,38.50,3230.36,26.75,10.64


## Use the ORDER BY clause

Now, imagine we were asked by our manager to figure out which 10 counties had the lowest birth count for 2016-2018. We could accomplish this by modifying our query to use the ORDER BY clause.

In [5]:
query = """
SELECT 
  *
FROM 
  `bigquery-public-data.sdoh_cdc_wonder_natality.county_natality` 
ORDER BY
  Births
LIMIT 
  10
"""

data = client.query(query).to_dataframe()
data

,Year,County_of_Residence,County_of_Residence_FIPS,Births,Ave_Age_of_Mother,Ave_OE_Gestational_Age_Wks,Ave_LMP_Gestational_Age_Wks,Ave_Birth_Weight_gms,Ave_Pre_pregnancy_BMI,Ave_Number_of_Prenatal_Wks
0,2018-01-01,"Tompkins County, NY",36109,735,30.49,38.80,38.96,3330.20,26.35,11.97
1,2018-01-01,"Unidentified Counties, HI",15999,749,29.78,38.61,38.92,3202.40,25.73,11.71
2,2017-01-01,"Tompkins County, NY",36109,767,30.27,38.89,39.07,3357.22,25.94,11.76
3,2016-01-01,"Tompkins County, NY",36109,787,30.49,38.92,39.04,3369.37,25.94,11.88
4,2018-01-01,"Unidentified Counties, MA",25999,802,30.54,39.02,39.05,3402.09,26.37,13.41
5,2016-01-01,"Unidentified Counties, HI",15999,845,29.18,38.59,38.79,3220.83,25.83,11.28
6,2017-01-01,"Washington County, RI",44009,852,30.59,38.75,38.99,3372.16,26.02,13.29
7,2017-01-01,"Unidentified Counties, HI",15999,857,29.29,38.56,38.86,3232.84,26.04,11.17
8,2017-01-01,"Unidentified Counties, MA",25999,866,30.53,38.97,38.98,3375.15,26.30,12.99
9,2018-01-01,"Fayette County, GA",13113,867,30.32,38.25,38.17,3233.16,26.56,12.01


We noticed that Tompkins County, NY, had just 735 births in 2018—the lowest birth count of any county in the US between 2016-2018.

### Use DESC to reverse sorting order

Now, we can also modify the query to sort in the other direction, returning the top 10 counties with the highest yearly birth counts between 2016-2018. 

In [7]:
query = """
SELECT 
  *
FROM 
  `bigquery-public-data.sdoh_cdc_wonder_natality.county_natality` 
ORDER BY
  Births
DESC
LIMIT 
  10
"""

data = client.query(query).to_dataframe()
data

,Year,County_of_Residence,County_of_Residence_FIPS,Births,Ave_Age_of_Mother,Ave_OE_Gestational_Age_Wks,Ave_LMP_Gestational_Age_Wks,Ave_Birth_Weight_gms,Ave_Pre_pregnancy_BMI,Ave_Number_of_Prenatal_Wks
0,2016-01-01,"Los Angeles County, CA",06037,123092,29.90,38.56,38.66,3260.01,26.16,11.99
1,2017-01-01,"Los Angeles County, CA",06037,116950,30.11,38.56,38.64,3259.04,26.38,12.12
2,2018-01-01,"Los Angeles County, CA",06037,110271,30.40,38.54,38.63,3260.45,26.60,12.19
3,2016-01-01,"Harris County, TX",48201,72420,28.52,38.27,38.40,3223.41,26.82,9.72
4,2017-01-01,"Harris County, TX",48201,68422,28.66,38.24,38.34,3215.85,26.93,9.81
5,2018-01-01,"Harris County, TX",48201,67095,28.82,38.19,38.32,3203.57,27.27,9.86
6,2016-01-01,"Cook County, IL",17031,66779,29.64,38.42,38.56,3233.71,26.89,10.80
7,2017-01-01,"Cook County, IL",17031,64374,29.79,38.43,38.57,3233.49,27.01,10.86
8,2018-01-01,"Cook County, IL",17031,61797,30.05,38.38,38.51,3230.21,27.13,10.88
9,2016-01-01,"Unidentified Counties, TX",48999,59168,26.53,38.31,38.44,3246.56,27.59,10.43


From the 10 rows with the largest values in the Birth column, we see that Los Angeles County takes up the top three spots.

## Combine ORDER BY with WHERE clauses

Next, we continue to modify the query so that it returns the top 10 counties with the highest birth counts for 2018 only. To do this, we add a WHERE clause to the query that specifies only rows that have a Year value equal to 2018-01-01. 

In [8]:
query = """
SELECT 
  *
FROM 
  `bigquery-public-data.sdoh_cdc_wonder_natality.county_natality` 
WHERE 
  Year = '2018-01-01'
ORDER BY
  Births
DESC
LIMIT 
  10
"""

data = client.query(query).to_dataframe()
data

,Year,County_of_Residence,County_of_Residence_FIPS,Births,Ave_Age_of_Mother,Ave_OE_Gestational_Age_Wks,Ave_LMP_Gestational_Age_Wks,Ave_Birth_Weight_gms,Ave_Pre_pregnancy_BMI,Ave_Number_of_Prenatal_Wks
0,2018-01-01,"Los Angeles County, CA",06037,110271,30.40,38.54,38.63,3260.45,26.60,12.19
1,2018-01-01,"Harris County, TX",48201,67095,28.82,38.19,38.32,3203.57,27.27,9.86
2,2018-01-01,"Cook County, IL",17031,61797,30.05,38.38,38.51,3230.21,27.13,10.88
3,2018-01-01,"Unidentified Counties, TX",48999,56295,26.89,38.26,38.42,3247.91,27.96,10.52
4,2018-01-01,"Maricopa County, AZ",04013,51854,28.69,38.44,38.53,3281.98,27.06,11.09
5,2018-01-01,"Unidentified Counties, GA",13999,41000,26.97,38.17,38.38,3197.04,28.46,11.24
6,2018-01-01,"San Diego County, CA",06073,40070,30.31,38.64,38.69,3300.06,26.06,11.86
7,2018-01-01,"Kings County, NY",36047,38574,30.39,38.68,38.73,3236.79,25.04,10.96
8,2018-01-01,"Unidentified Counties, VA",51999,38308,28.39,38.50,38.69,3276.37,28.05,11.37
9,2018-01-01,"Dallas County, TX",48113,38182,28.48,38.40,38.57,3235.31,25.92,9.77


From the results, we can see that the Los Angeles County, CA still took the 1st position with 110,271 births in 2018.